internal/keysign/keysign.go

```diff
@@ -0,0 +1,129 @@
+package keysign
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	tsslib "github.com/vultisig/mobile-tss-lib/tss"
+
+	"github.com/vultisig/vultisigner/config"
+	"github.com/vultisig/vultisigner/internal/logging"
+	"github.com/vultisig/vultisigner/internal/tss"
+	"github.com/vultisig/vultisigner/internal/types"
+	"github.com/vultisig/vultisigner/relay"
+)
+
+func JoinKeySign(ks *types.KeysignRequest) ([]string, error) {
+	result := []string{}
+	keyFolder := config.AppConfig.Server.VaultsFilePath
+	serverURL := config.AppConfig.Relay.Server
+
+	server := relay.NewServer(serverURL)
+
+	// Let's register session here
+	if err := server.RegisterSession(ks.Session, ks.Key); err != nil {
+		return result, fmt.Errorf("failed to register session: %w", err)
+	}
+	// wait longer for keysign start
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	partiesJoined, err := server.WaitForSessionStart(ctx, ks.Session)
+	logging.Logger.WithFields(logrus.Fields{
+		"session":        ks.Session,
+		"parties_joined": partiesJoined,
+	}).Info("Session started")
+
+	if err != nil {
+		return result, fmt.Errorf("failed to wait for session start: %w", err)
+	}
+
+	localStateAccessor := &relay.LocalStateAccessorImp{
+		Key:    ks.Key,
+		Folder: keyFolder,
+	}
+	tssServerImp, err := tss.CreateTSSService(serverURL, ks.Session, ks.HexEncryptionKey, localStateAccessor)
+	if err != nil {
+		return result, fmt.Errorf("failed to create TSS service: %w", err)
+	}
+
+	for _, message := range ks.Messages {
+		var signatureEncoded string
+		for attempt := 0; attempt < 3; attempt++ {
+			signatureEncoded, err = keysignWithRetry(serverURL, ks, partiesJoined, tssServerImp, message)
+			if err == nil {
+				break
+			}
+		}
+		if err != nil {
+			return result, err
+		}
+		result = append(result, signatureEncoded)
+	}
+
+	if err := server.CompleteSession(ks.Session, ks.Key); err != nil {
+		logging.Logger.WithFields(logrus.Fields{
+			"session": ks.Session,
+			"error":   err,
+		}).Error("Failed to complete session")
+	}
+
+	if err := server.EndSession(ks.Session); err != nil {
```

Comment:
We probably should not end the session , in case other party didn't get the last messages